src/urlencoded.rs

```diff
@@ -1,11 +1,137 @@
-use std::collections::HashMap;
-use url::form_urlencoded;
 use groupable::Groupable;
+use hyper::uri::RequestUri;
+use hyper::uri::RequestUri::{Star, AbsoluteUri, AbsolutePath, Authority};
+use plugin::{Plugin, Pluggable};
+use request::Request;
+use status::StatusCode;
+use std::collections::HashMap;
+use std::io::{Error, Read};
+use typemap::Key;
+use url::{form_urlencoded, UrlParser};
+
+type QueryStore = HashMap<String, Vec<String>>;
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct Query(QueryStore);
+
+impl Query {
+    /// Retrieves the first value from the query for `key`, or `None` if not present.
+    ///
+    /// # Notes
+    /// There may be multiple values per key, if all of the values for a given
+    /// `key` are required, then use `all`.
+    //FIXME: Implement via Indexing whenever IndexGet is supported
+    pub fn get(&self, key: &str) -> Option<&str> {
+        self.0.get(key).and_then(|v| v.first().map(|s| &**s))
+    }
 
-pub fn parse (encoded_string : &str) -> HashMap<String, Vec<String>> {
+    /// Retrieve all values from the query for `key`, or `None` if none are present.
+    pub fn all(&self, key: &str) -> Option<&[String]> {
+        self.0.get(key).map(|v| &**v)
+    }
+}
+
+fn parse (encoded_string : &str) -> QueryStore {
     form_urlencoded::parse(encoded_string.as_bytes()).into_iter().group()
 }
 
+fn parse_uri(origin: &RequestUri) -> Query {
+    let f = |query: Option<&String>| query.map(|q| parse(&*q));
+
+    let result = match *origin {
+        AbsoluteUri(ref url) => f(url.query.as_ref()),
+        AbsolutePath(ref s) => UrlParser::new().parse_path(&*s)
+                                                // FIXME: If this fails to parse,
+                                                // then it really shouldn't have
+                                                // reached here.
+                                               .ok()
+                                               .and_then(|(_, query, _)| f(query.as_ref())),
+        Star | Authority(..) => None
+    };
+
+    Query(result.unwrap_or_else(|| HashMap::new()))
+}
+
+struct FormBodyParser;
+
+impl Key for FormBodyParser {
+    type Value = Query;
+}
+
+impl<'mw, 'conn, D> Plugin<Request<'mw, 'conn, D>> for FormBodyParser {
+    type Error = Error;
+
+    fn eval(req: &mut Request<D>) -> Result<Query, Error> {
+        let mut buf = String::new();
+        try!(req.origin.read_to_string(&mut buf));
```

Comment:
This and the json equivalent should likely depend on a `Plugin` which just reads and caches the body (as `Vec<u8>`), else if a user tries to parse json, fails, then parses the form, I think it would fail. (Ultimately this is a failure of the current lack of robust api on `Request`).

It would be good to capture this in an integration test here: https://github.com/nickel-org/nickel.rs/tree/master/tests/examples